// @stryker-mutator/core/src/logging/log-configurator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogConfigurator = void 0;
const log4js = require("log4js");
const net_utils_1 = require("../utils/net-utils");
const log_utils_1 = require("./log-utils");
const layouts = {
    color: {
        pattern: '%[%r (%z) %p %c%] %m',
        type: 'pattern',
    },
    noColor: {
        pattern: '%r (%z) %p %c %m',
        type: 'pattern',
    },
};
const LOG_FILE_NAME = 'stryker.log';
class LogConfigurator {
    static createMainProcessAppenders(consoleLogLevel, fileLogLevel, allowConsoleColors) {
        // Add the custom "multiAppender": https://log4js-node.github.io/log4js-node/appenders.html#other-appenders
        const multiAppender = { type: require.resolve('./multi-appender'), appenders: ["filteredConsoleLevel" /* FilteredConsoleLevel */] };
        const consoleLayout = allowConsoleColors ? layouts.color : layouts.noColor;
        let allAppenders = {
            ["console" /* Console */]: { type: 'stdout', layout: consoleLayout },
            // Exclude messages like: "ERROR log4js A worker log process hung up unexpectedly" #1245
            ["filteredConsoleCategory" /* FilteredConsoleCategory */]: { type: 'categoryFilter', appender: "console" /* Console */, exclude: 'log4js' },
            ["filteredConsoleLevel" /* FilteredConsoleLevel */]: { type: 'logLevelFilter', appender: "filteredConsoleCategory" /* FilteredConsoleCategory */, level: consoleLogLevel },
            ["all" /* All */]: multiAppender,
        };
        // only add file if it is needed. Otherwise log4js will create the file directly, pretty annoying.
        if (fileLogLevel.toUpperCase() !== "off" /* Off */.toUpperCase()) {
            const fileAppender = { type: 'file', filename: LOG_FILE_NAME, layout: layouts.noColor };
            const filteredFileAppender = { type: 'logLevelFilter', appender: 'file', level: fileLogLevel };
            // Don't simply add the appenders, instead actually make sure they are ordinal "before" the others.
            // See https://github.com/log4js-node/log4js-node/issues/746
            allAppenders = Object.assign({ ["file" /* File */]: fileAppender, ["filteredFile" /* FilteredFile */]: filteredFileAppender }, allAppenders);
            multiAppender.appenders.push("filteredFile" /* FilteredFile */);
        }
        return allAppenders;
    }
    static createLog4jsConfig(defaultLogLevel, appenders) {
        return {
            appenders,
            categories: {
                default: {
                    appenders: ["all" /* All */],
                    level: defaultLogLevel,
                },
            },
        };
    }
    /**
     * Configure logging for the master process. Either call this method or `configureChildProcess` before any `getLogger` calls.
     * @param consoleLogLevel The log level to configure for the console
     * @param fileLogLevel The log level to configure for the "stryker.log" file
     */
    static configureMainProcess(consoleLogLevel = "info" /* Information */, fileLogLevel = "off" /* Off */, allowConsoleColors = true) {
        const appenders = this.createMainProcessAppenders(consoleLogLevel, fileLogLevel, allowConsoleColors);
        log4js.configure(this.createLog4jsConfig(log_utils_1.minLevel(consoleLogLevel, fileLogLevel), appenders));
    }
    /**
     * Configure the logging for the server. Includes the master configuration.
     * This method should only be called ONCE, as it starts the log4js server to listen for log events.
     * It returns the logging client context that should be used to configure the child processes.
     *
     * @param consoleLogLevel the console log level
     * @param fileLogLevel the file log level
     * @returns the context
     */
    static async configureLoggingServer(consoleLogLevel, fileLogLevel, allowConsoleColors) {
        const loggerPort = await net_utils_1.getFreePort();
        // Include the appenders for the main Stryker process, as log4js has only one single `configure` method.
        const appenders = this.createMainProcessAppenders(consoleLogLevel, fileLogLevel, allowConsoleColors);
        const multiProcessAppender = {
            appender: "all" /* All */,
            loggerPort,
            mode: 'master',
            type: 'multiprocess',
        };
        appenders["server" /* Server */] = multiProcessAppender;
        const defaultLogLevel = log_utils_1.minLevel(consoleLogLevel, fileLogLevel);
        log4js.configure(this.createLog4jsConfig(defaultLogLevel, appenders));
        const context = {
            level: defaultLogLevel,
            port: loggerPort,
        };
        return context;
    }
    /**
     * Configures the logging for a worker process. Sends all logging to the master process.
     * Either call this method or `configureMainProcess` before any `getLogger` calls.
     * @param context the logging client context used to configure the logging client
     */
    static configureChildProcess(context) {
        const clientAppender = { type: 'multiprocess', mode: 'worker', loggerPort: context.port };
        const appenders = { ["all" /* All */]: clientAppender };
        log4js.configure(this.createLog4jsConfig(context.level, appenders));
    }
    static shutdown() {
        return new Promise((res, rej) => {
            log4js.shutdown((err) => {
                if (err) {
                    rej(err);
                }
                else {
                    res();
                }
            });
        });
    }
}
exports.LogConfigurator = LogConfigurator;
