// @stryker-mutator/core/src/utils/file-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.findNodeModules = exports.symlinkJunction = exports.writeFile = exports.importModule = exports.cleanFolder = exports.deleteDir = exports.glob = exports.MAX_CONCURRENT_FILE_IO = void 0;
const path = require("path");
const fs_1 = require("fs");
const util_1 = require("util");
const nodeGlob = require("glob");
const mkdirp = require("mkdirp");
const rimraf = require("rimraf");
exports.MAX_CONCURRENT_FILE_IO = 256;
function glob(expression) {
    return new Promise((resolve, reject) => {
        nodeGlob(expression, { nodir: true }, (error, matches) => {
            error ? reject(error) : resolve(matches);
        });
    });
}
exports.glob = glob;
exports.deleteDir = util_1.promisify(rimraf);
async function cleanFolder(folderName) {
    try {
        await fs_1.promises.lstat(folderName);
        await exports.deleteDir(folderName);
        return mkdirp.sync(folderName);
    }
    catch (e) {
        return mkdirp.sync(folderName);
    }
}
exports.cleanFolder = cleanFolder;
/**
 * Wrapper around the 'require' function (for testability)
 */
function importModule(moduleName) {
    return require(moduleName);
}
exports.importModule = importModule;
/**
 * Writes data to a specified file.
 * @param fileName The path to the file.
 * @param data The content of the file.
 * @returns A promise to eventually save the file.
 */
function writeFile(fileName, data) {
    if (Buffer.isBuffer(data)) {
        return fs_1.promises.writeFile(fileName, data);
    }
    else {
        return fs_1.promises.writeFile(fileName, data, 'utf8');
    }
}
exports.writeFile = writeFile;
/**
 * Creates a symlink at `from` that points to `to`
 * @param to The thing you want to point to
 * @param from The thing you want to point from
 */
function symlinkJunction(to, from) {
    return fs_1.promises.symlink(to, from, 'junction');
}
exports.symlinkJunction = symlinkJunction;
/**
 * Looks for the node_modules folder from basePath up to root.
 * returns the first occurrence of the node_modules, or null of none could be found.
 * @param basePath starting point
 */
async function findNodeModules(basePath) {
    basePath = path.resolve(basePath);
    const nodeModules = path.resolve(basePath, 'node_modules');
    try {
        await fs_1.promises.stat(nodeModules);
        return nodeModules;
    }
    catch (e) {
        const parent = path.dirname(basePath);
        if (parent === basePath) {
            return null;
        }
        else {
            return findNodeModules(path.dirname(basePath));
        }
    }
}
exports.findNodeModules = findNodeModules;
