// @stryker-mutator/core/src/initializer/npm-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const plugin_1 = require("@stryker-mutator/api/plugin");
const util_1 = require("@stryker-mutator/util");
const _1 = require(".");
const getName = (packageName) => {
    return packageName.replace('@stryker-mutator/', '').replace('stryker-', '').split('-')[0];
};
const mapSearchResultToPromptOption = (searchResults) => searchResults.results.map((result) => ({
    name: getName(result.package.name),
    pkg: result.package,
}));
const handleResult = (from) => (response) => {
    if (response.statusCode === 200 && response.result) {
        return response.result;
    }
    else {
        throw new Error(`Path ${from} resulted in http status code: ${response.statusCode}.`);
    }
};
class NpmClient {
    constructor(log, searchClient, packageClient) {
        this.log = log;
        this.searchClient = searchClient;
        this.packageClient = packageClient;
    }
    getTestRunnerOptions() {
        return this.search('/v2/search?q=keywords:@stryker-mutator/test-runner-plugin').then(mapSearchResultToPromptOption);
    }
    getTestReporterOptions() {
        return this.search('/v2/search?q=keywords:@stryker-mutator/reporter-plugin').then(mapSearchResultToPromptOption);
    }
    getAdditionalConfig(pkg) {
        const path = `/${pkg.name}@${pkg.version}/package.json`;
        return this.packageClient
            .get(path)
            .then(handleResult(path))
            .then((pkg) => pkg.initStrykerConfig || {})
            .catch((err) => {
            this.log.warn(`Could not fetch additional initialization config for dependency ${pkg.name}. You might need to configure it manually`, err);
            return {};
        });
    }
    search(path) {
        this.log.debug(`Searching: ${path}`);
        return this.searchClient
            .get(path)
            .then(handleResult(path))
            .catch((err) => {
            this.log.error(`Unable to reach npms.io (for query ${path}). Please check your internet connection.`, util_1.errorToString(err));
            const result = {
                results: [],
                total: 0,
            };
            return result;
        });
    }
}
exports.default = NpmClient;
NpmClient.inject = plugin_1.tokens(plugin_1.commonTokens.logger, _1.initializerTokens.restClientNpmSearch, _1.initializerTokens.restClientNpm);
