// @stryker-mutator/core/src/initializer/stryker-initializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const child = require("child_process");
const fs_1 = require("fs");
const plugin_1 = require("@stryker-mutator/api/plugin");
const util_1 = require("@stryker-mutator/util");
const _1 = require(".");
class StrykerInitializer {
    constructor(log, out, client, strykerPresets, configWriter, gitignoreWriter, inquirer) {
        this.log = log;
        this.out = out;
        this.client = client;
        this.strykerPresets = strykerPresets;
        this.configWriter = configWriter;
        this.gitignoreWriter = gitignoreWriter;
        this.inquirer = inquirer;
    }
    /**
     * Runs the initializer will prompt the user for questions about his setup. After that, install plugins and configure Stryker.
     * @function
     */
    async initialize() {
        this.configWriter.guardForExistingConfig();
        this.patchProxies();
        const selectedPreset = await this.selectPreset();
        let configFileName;
        if (selectedPreset) {
            configFileName = await this.initiatePreset(this.configWriter, selectedPreset);
        }
        else {
            configFileName = await this.initiateCustom(this.configWriter);
        }
        await this.gitignoreWriter.addStrykerTempFolder();
        this.out(`Done configuring stryker. Please review "${configFileName}", you might need to configure your test runner correctly.`);
        this.out("Let's kill some mutants with this command: `stryker run`");
    }
    /**
     * The typed rest client works only with the specific HTTP_PROXY and HTTPS_PROXY env settings.
     * Let's make sure they are available.
     */
    patchProxies() {
        const copyEnvVariable = (from, to) => {
            if (process.env[from] && !process.env[to]) {
                process.env[to] = process.env[from];
            }
        };
        copyEnvVariable('http_proxy', 'HTTP_PROXY');
        copyEnvVariable('https_proxy', 'HTTPS_PROXY');
    }
    async selectPreset() {
        const presetOptions = this.strykerPresets;
        if (presetOptions.length) {
            this.log.debug(`Found presets: ${JSON.stringify(presetOptions)}`);
            return this.inquirer.promptPresets(presetOptions);
        }
        else {
            this.log.debug('No presets have been configured, reverting to custom configuration');
            return undefined;
        }
    }
    async initiatePreset(configWriter, selectedPreset) {
        const presetConfig = await selectedPreset.createConfig();
        const isJsonSelected = await this.selectJsonConfigType();
        const configFileName = await configWriter.writePreset(presetConfig, isJsonSelected);
        if (presetConfig.additionalConfigFiles) {
            await Promise.all(presetConfig.additionalConfigFiles.map(({ name, content }) => fs_1.promises.writeFile(name, content)));
        }
        const selectedPackageManager = await this.selectPackageManager();
        this.installNpmDependencies(presetConfig.dependencies, selectedPackageManager);
        return configFileName;
    }
    async initiateCustom(configWriter) {
        const selectedTestRunner = await this.selectTestRunner();
        const selectedReporters = await this.selectReporters();
        const selectedPackageManager = await this.selectPackageManager();
        const isJsonSelected = await this.selectJsonConfigType();
        const npmDependencies = this.getSelectedNpmDependencies([selectedTestRunner].concat(selectedReporters));
        const configFileName = await configWriter.write(selectedTestRunner, selectedReporters, selectedPackageManager, await this.fetchAdditionalConfig(npmDependencies), isJsonSelected);
        this.installNpmDependencies(npmDependencies.map((pkg) => pkg.name), selectedPackageManager);
        return configFileName;
    }
    async selectTestRunner() {
        const testRunnerOptions = await this.client.getTestRunnerOptions();
        this.log.debug(`Found test runners: ${JSON.stringify(testRunnerOptions)}`);
        return this.inquirer.promptTestRunners(testRunnerOptions);
    }
    async selectReporters() {
        let reporterOptions;
        reporterOptions = await this.client.getTestReporterOptions();
        reporterOptions.push({
            name: 'html',
            pkg: null,
        }, {
            name: 'clear-text',
            pkg: null,
        }, {
            name: 'progress',
            pkg: null,
        }, {
            name: 'dashboard',
            pkg: null,
        });
        return this.inquirer.promptReporters(reporterOptions);
    }
    async selectPackageManager() {
        return this.inquirer.promptPackageManager([
            {
                name: "npm" /* Npm */,
                pkg: null,
            },
            {
                name: "yarn" /* Yarn */,
                pkg: null,
            },
        ]);
    }
    async selectJsonConfigType() {
        return this.inquirer.promptJsonConfigType();
    }
    getSelectedNpmDependencies(selectedOptions) {
        return selectedOptions
            .filter(util_1.notEmpty)
            .map((option) => option.pkg)
            .filter(util_1.notEmpty);
    }
    /**
     * Install the npm packages
     * @function
     */
    installNpmDependencies(dependencies, selectedOption) {
        if (dependencies.length === 0) {
            return;
        }
        const dependencyArg = dependencies.join(' ');
        this.out('Installing NPM dependencies...');
        const cmd = selectedOption.name === "npm" /* Npm */ ? `npm i --save-dev ${dependencyArg}` : `yarn add ${dependencyArg} --dev`;
        this.out(cmd);
        try {
            child.execSync(cmd, { stdio: [0, 1, 2] });
        }
        catch (_) {
            this.out(`An error occurred during installation, please try it yourself: "${cmd}"`);
        }
    }
    async fetchAdditionalConfig(dependencies) {
        return (await Promise.all(dependencies.map((dep) => this.client.getAdditionalConfig(dep)))).filter(util_1.notEmpty);
    }
}
exports.default = StrykerInitializer;
StrykerInitializer.inject = plugin_1.tokens(plugin_1.commonTokens.logger, _1.initializerTokens.out, _1.initializerTokens.npmClient, _1.initializerTokens.strykerPresets, _1.initializerTokens.configWriter, _1.initializerTokens.gitignoreWriter, _1.initializerTokens.inquirer);
