// @stryker-mutator/core/src/child-proxy/child-process-proxy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const child_process_1 = require("child_process");
const os = require("os");
const core_1 = require("@stryker-mutator/api/core");
const util_1 = require("@stryker-mutator/util");
const log4js_1 = require("log4js");
const object_utils_1 = require("../utils/object-utils");
const string_builder_1 = require("../utils/string-builder");
const child_process_crashed_error_1 = require("./child-process-crashed-error");
const message_protocol_1 = require("./message-protocol");
const out_of_memory_error_1 = require("./out-of-memory-error");
const BROKEN_PIPE_ERROR_CODE = 'EPIPE';
const IPC_CHANNEL_CLOSED_ERROR_CODE = 'ERR_IPC_CHANNEL_CLOSED';
const TIMEOUT_FOR_DISPOSE = 2000;
class ChildProcessProxy {
    constructor(requirePath, requireName, loggingContext, options, additionalInjectableValues, workingDirectory) {
        this.workerTasks = [];
        this.log = log4js_1.getLogger(ChildProcessProxy.name);
        this.stdoutBuilder = new string_builder_1.default();
        this.stderrBuilder = new string_builder_1.default();
        this.isDisposed = false;
        this.handleUnexpectedExit = (code, signal) => {
            this.isDisposed = true;
            const output = string_builder_1.default.concat(this.stderrBuilder, this.stdoutBuilder);
            if (processOutOfMemory()) {
                this.currentError = new out_of_memory_error_1.default(this.worker.pid, code);
                this.log.warn(`Child process [pid ${this.currentError.pid}] ran out of memory. Stdout and stderr are logged on debug level.`);
                this.log.debug(stdoutAndStderr());
            }
            else {
                this.currentError = new child_process_crashed_error_1.default(this.worker.pid, `Child process [pid ${this.worker.pid}] exited unexpectedly with exit code ${code} (${signal || 'without signal'}). ${stdoutAndStderr()}`, code, signal);
                this.log.warn(this.currentError.message, this.currentError);
            }
            this.reportError(this.currentError);
            function processOutOfMemory() {
                return output.includes('JavaScript heap out of memory');
            }
            function stdoutAndStderr() {
                if (output.length) {
                    return `Last part of stdout and stderr was:${os.EOL}${object_utils_1.padLeft(output)}`;
                }
                else {
                    return 'Stdout and stderr were empty.';
                }
            }
        };
        this.handleError = (error) => {
            if (this.innerProcessIsCrashed(error)) {
                this.log.warn(`Child process [pid ${this.worker.pid}] has crashed. See other warning messages for more info.`, error);
                this.reportError(new child_process_crashed_error_1.default(this.worker.pid, `Child process [pid ${this.worker.pid}] has crashed`, undefined, undefined, error));
            }
            else {
                this.reportError(error);
            }
        };
        this.worker = child_process_1.fork(require.resolve('./child-process-proxy-worker'), [message_protocol_1.autoStart], { silent: true, execArgv: [] });
        this.initTask = new util_1.Task();
        this.log.debug('Starting %s in child process %s', requirePath, this.worker.pid);
        this.send({
            additionalInjectableValues,
            kind: message_protocol_1.WorkerMessageKind.Init,
            loggingContext,
            options,
            requireName,
            requirePath,
            workingDirectory,
        });
        this.listenForMessages();
        this.listenToStdoutAndStderr();
        // Listen to `close`, not `exit`, see https://github.com/stryker-mutator/stryker/issues/1634
        this.worker.on('close', this.handleUnexpectedExit);
        this.worker.on('error', this.handleError);
        this.proxy = this.initProxy();
    }
    /**
     * @description Creates a proxy where each function of the object created using the constructorFunction arg is ran inside of a child process
     */
    static create(requirePath, loggingContext, options, additionalInjectableValues, workingDirectory, injectableClass) {
        return new ChildProcessProxy(requirePath, injectableClass.name, loggingContext, options, additionalInjectableValues, workingDirectory);
    }
    send(message) {
        this.worker.send(object_utils_1.serialize(message, [core_1.File]));
    }
    initProxy() {
        // This proxy is a genuine javascript `Proxy` class
        // More info: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy
        const self = this;
        return new Proxy({}, {
            get(_, propertyKey) {
                if (typeof propertyKey === 'string') {
                    return self.forward(propertyKey);
                }
                else {
                    return undefined;
                }
            },
        });
    }
    forward(methodName) {
        return async (...args) => {
            if (this.currentError) {
                return Promise.reject(this.currentError);
            }
            else {
                const workerTask = new util_1.Task();
                const correlationId = this.workerTasks.push(workerTask) - 1;
                this.initTask.promise.then(() => {
                    this.send({
                        args,
                        correlationId,
                        kind: message_protocol_1.WorkerMessageKind.Call,
                        methodName,
                    });
                });
                return workerTask.promise;
            }
        };
    }
    listenForMessages() {
        this.worker.on('message', (serializedMessage) => {
            const message = object_utils_1.deserialize(serializedMessage, [core_1.File]);
            switch (message.kind) {
                case message_protocol_1.ParentMessageKind.Initialized:
                    this.initTask.resolve(undefined);
                    break;
                case message_protocol_1.ParentMessageKind.Result:
                    this.workerTasks[message.correlationId].resolve(message.result);
                    delete this.workerTasks[message.correlationId];
                    break;
                case message_protocol_1.ParentMessageKind.Rejection:
                    this.workerTasks[message.correlationId].reject(new Error(message.error));
                    delete this.workerTasks[message.correlationId];
                    break;
                case message_protocol_1.ParentMessageKind.DisposeCompleted:
                    if (this.disposeTask) {
                        this.disposeTask.resolve(undefined);
                    }
                    break;
                default:
                    this.logUnidentifiedMessage(message);
                    break;
            }
        });
    }
    listenToStdoutAndStderr() {
        const handleData = (builder) => (data) => {
            const output = data.toString();
            builder.append(output);
            if (this.log.isTraceEnabled()) {
                this.log.trace(output);
            }
        };
        if (this.worker.stdout) {
            this.worker.stdout.on('data', handleData(this.stdoutBuilder));
        }
        if (this.worker.stderr) {
            this.worker.stderr.on('data', handleData(this.stderrBuilder));
        }
    }
    get stdout() {
        return this.stdoutBuilder.toString();
    }
    reportError(error) {
        this.workerTasks.filter((task) => !task.isCompleted).forEach((task) => task.reject(error));
    }
    innerProcessIsCrashed(error) {
        return util_1.isErrnoException(error) && (error.code === BROKEN_PIPE_ERROR_CODE || error.code === IPC_CHANNEL_CLOSED_ERROR_CODE);
    }
    async dispose() {
        if (!this.isDisposed) {
            this.worker.removeListener('close', this.handleUnexpectedExit);
            this.isDisposed = true;
            this.log.debug('Disposing of worker process %s', this.worker.pid);
            this.disposeTask = new util_1.ExpirableTask(TIMEOUT_FOR_DISPOSE);
            this.send({ kind: message_protocol_1.WorkerMessageKind.Dispose });
            try {
                await this.disposeTask.promise;
            }
            finally {
                this.log.debug('Kill %s', this.worker.pid);
                await object_utils_1.kill(this.worker.pid);
            }
        }
    }
    logUnidentifiedMessage(message) {
        this.log.error(`Received unidentified message ${message}`);
    }
}
exports.default = ChildProcessProxy;
