// @stryker-mutator/core/src/child-proxy/message-protocol.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.autoStart = exports.ParentMessageKind = exports.WorkerMessageKind = void 0;
var WorkerMessageKind;
(function (WorkerMessageKind) {
    WorkerMessageKind[WorkerMessageKind["Init"] = 0] = "Init";
    WorkerMessageKind[WorkerMessageKind["Call"] = 1] = "Call";
    WorkerMessageKind[WorkerMessageKind["Dispose"] = 2] = "Dispose";
})(WorkerMessageKind = exports.WorkerMessageKind || (exports.WorkerMessageKind = {}));
var ParentMessageKind;
(function (ParentMessageKind) {
    ParentMessageKind[ParentMessageKind["Initialized"] = 0] = "Initialized";
    ParentMessageKind[ParentMessageKind["Result"] = 1] = "Result";
    ParentMessageKind[ParentMessageKind["Rejection"] = 2] = "Rejection";
    ParentMessageKind[ParentMessageKind["DisposeCompleted"] = 3] = "DisposeCompleted";
})(ParentMessageKind = exports.ParentMessageKind || (exports.ParentMessageKind = {}));
// Make this an unlikely command line argument
// (prevents incidental start of child process)
exports.autoStart = 'childProcessAutoStart12937129s7d';
