// @stryker-mutator/instrumenter/dist/src/parsers/ts-parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parse = void 0;
const core_1 = require("@babel/core");
const core_2 = require("@babel/core");
const syntax_1 = require("../syntax");
/**
 * See https://babeljs.io/docs/en/babel-preset-typescript
 * @param text The text to parse
 * @param fileName The name of the file
 */
async function parse(text, fileName) {
    const isTSX = fileName.endsWith('x');
    const ast = await core_2.parseAsync(text, {
        filename: fileName,
        parserOpts: {
            ranges: true,
        },
        configFile: false,
        babelrc: false,
        presets: [[require.resolve('@babel/preset-typescript'), { isTSX, allExtensions: true }]],
        plugins: [
            require.resolve('@babel/plugin-proposal-class-properties'),
            require.resolve('@babel/plugin-proposal-private-methods'),
            [require.resolve('@babel/plugin-proposal-decorators'), { legacy: true }],
        ],
    });
    if (core_1.types.isProgram(ast)) {
        throw new Error(`Expected ${fileName} to contain a babel.types.file, but was a program`);
    }
    return {
        originFileName: fileName,
        rawContent: text,
        format: syntax_1.AstFormat.TS,
        root: ast,
    };
}
exports.parse = parse;
