// @stryker-mutator/api/src/report/mutant-status.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var MutantStatus;
(function (MutantStatus) {
    /**
     * The status of a survived mutant, because it was not covered by any test.
     */
    MutantStatus[MutantStatus["NoCoverage"] = 0] = "NoCoverage";
    /**
     * The status of a killed mutant.
     */
    MutantStatus[MutantStatus["Killed"] = 1] = "Killed";
    /**
     * The status of a survived mutant.
     */
    MutantStatus[MutantStatus["Survived"] = 2] = "Survived";
    /**
     * The status of a timed out mutant.
     */
    MutantStatus[MutantStatus["TimedOut"] = 3] = "TimedOut";
    /**
     * The status of a mutant of which the tests resulted in a runtime error.
     * @example
     * ```javascript
     * const fs = require('f' - 's'); // mutated code
     * ```
     */
    MutantStatus[MutantStatus["RuntimeError"] = 4] = "RuntimeError";
    /**
     * The status of a mutant which could not be compiled.
     * @example
     * ```typescript
     * const foo = 'foo' - 'bar'; // mutated code
     * ```
     */
    MutantStatus[MutantStatus["CompileError"] = 5] = "CompileError";
    /**
     * The status of a mutant that is ignored. For example, by user configuration.
     */
    MutantStatus[MutantStatus["Ignored"] = 6] = "Ignored";
})(MutantStatus || (MutantStatus = {}));
exports.default = MutantStatus;
