// @stryker-mutator/core/src/reporters/event-recorder-reporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const fs_1 = require("fs");
const plugin_1 = require("@stryker-mutator/api/plugin");
const file_utils_1 = require("../utils/file-utils");
class EventRecorderReporter {
    constructor(log, options) {
        this.log = log;
        this.options = options;
        this.allWork = [];
        this.index = 0;
        this.createBaseFolderTask = file_utils_1.cleanFolder(this.options.eventReporter.baseDir);
    }
    writeToFile(methodName, data) {
        const filename = path.join(this.options.eventReporter.baseDir, `${this.format(this.index++)}-${methodName}.json`);
        this.log.debug(`Writing event ${methodName} to file ${filename}`);
        return fs_1.promises.writeFile(filename, JSON.stringify(data), { encoding: 'utf8' });
    }
    format(input) {
        let str = input.toString();
        for (let i = 10000; i > 1; i = i / 10) {
            if (i > input) {
                str = '0' + str;
            }
        }
        return str;
    }
    work(eventName, data) {
        this.allWork.push(this.createBaseFolderTask.then(() => this.writeToFile(eventName, data)));
    }
    onSourceFileRead(file) {
        this.work('onSourceFileRead', file);
    }
    onAllSourceFilesRead(files) {
        this.work('onAllSourceFilesRead', files);
    }
    onAllMutantsMatchedWithTests(results) {
        this.work('onAllMutantsMatchedWithTests', results);
    }
    onMutantTested(result) {
        this.work('onMutantTested', result);
    }
    onMutationTestReportReady(report) {
        this.work('onMutationTestReportReady', report);
    }
    onAllMutantsTested(results) {
        this.work('onAllMutantsTested', results);
    }
    async wrapUp() {
        await this.createBaseFolderTask;
        return Promise.all(this.allWork);
    }
}
exports.default = EventRecorderReporter;
EventRecorderReporter.inject = plugin_1.tokens(plugin_1.commonTokens.logger, plugin_1.commonTokens.options);
