// @stryker-mutator/jest-runner/src/jest-test-adapters/jest-greater-than-25-adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const plugin_1 = require("@stryker-mutator/api/plugin");
const jest_wrapper_1 = require("../utils/jest-wrapper");
class JestGreaterThan25Adapter {
    constructor(log) {
        this.log = log;
    }
    async run(jestConfig, projectRoot, fileNameUnderTest) {
        jestConfig.reporters = [];
        const config = JSON.stringify(jestConfig);
        this.log.trace(`Invoking Jest with config ${config}`);
        if (fileNameUnderTest) {
            this.log.trace(`Only running tests related to ${fileNameUnderTest}`);
        }
        const result = await jest_wrapper_1.jestWrapper.runCLI({
            $0: 'stryker',
            _: fileNameUnderTest ? [fileNameUnderTest] : [],
            findRelatedTests: !!fileNameUnderTest,
            config,
            runInBand: true,
            silent: true,
        }, [projectRoot]);
        return result;
    }
}
exports.default = JestGreaterThan25Adapter;
JestGreaterThan25Adapter.inject = plugin_1.tokens(plugin_1.commonTokens.logger);
