// @stryker-mutator/core/src/reporters/dashboard-reporter/dashboard-reporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const plugin_1 = require("@stryker-mutator/api/plugin");
const mutation_testing_metrics_1 = require("mutation-testing-metrics");
const tokens_1 = require("./tokens");
class DashboardReporter {
    constructor(log, dashboardReporterClient, options, ciProvider) {
        this.log = log;
        this.dashboardReporterClient = dashboardReporterClient;
        this.options = options;
        this.ciProvider = ciProvider;
    }
    onMutationTestReportReady(result) {
        this.onGoingWork = (async () => {
            const { projectName, version, moduleName } = this.getContextFromEnvironment();
            if (projectName && version) {
                await this.update(this.toReport(result), projectName, version, moduleName);
            }
            else {
                this.log.info('The report was not send to the dashboard. The dashboard.project and/or dashboard.version values were missing and not detected to be running on a build server.');
            }
        })();
    }
    async wrapUp() {
        await this.onGoingWork;
    }
    toReport(result) {
        if (this.options.dashboard.reportType === "full" /* Full */) {
            return result;
        }
        else {
            return {
                mutationScore: mutation_testing_metrics_1.calculateMetrics(result.files).metrics.mutationScore,
            };
        }
    }
    async update(report, projectName, version, moduleName) {
        try {
            const href = await this.dashboardReporterClient.updateReport({
                report,
                moduleName,
                projectName,
                version: version,
            });
            this.log.info('Report available at: %s', href);
        }
        catch (err) {
            this.log.error('Could not upload report.', err);
        }
    }
    getContextFromEnvironment() {
        return {
            moduleName: this.options.dashboard.module,
            projectName: this.options.dashboard.project || (this.ciProvider && this.ciProvider.determineProject()),
            version: this.options.dashboard.version || (this.ciProvider && this.ciProvider.determineVersion()),
        };
    }
}
exports.default = DashboardReporter;
DashboardReporter.inject = plugin_1.tokens(plugin_1.commonTokens.logger, tokens_1.dashboardReporterTokens.dashboardReporterClient, plugin_1.commonTokens.options, tokens_1.dashboardReporterTokens.ciProvider);
