// @stryker-mutator/core/src/di/core-tokens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.optionsValidator = exports.validationSchema = exports.pluginCreatorTestRunner = exports.pluginCreatorChecker = exports.pluginCreatorReporter = exports.pluginDescriptors = exports.pluginKind = exports.reporter = exports.testRunnerConcurrencyTokens = exports.testRunnerPool = exports.testRunnerFactory = exports.concurrencyTokenProvider = exports.sandboxWorkingDirectory = exports.sandbox = exports.mutationTestReportHelper = exports.loggingContext = exports.timeOverheadMS = exports.timer = exports.temporaryDirectory = exports.mutantTranspileScheduler = exports.mutantsWithTestCoverage = exports.mutants = exports.files = exports.transpiledFiles = exports.dryRunResult = exports.inputFiles = exports.configOptionsApplier = exports.configReader = exports.cliOptions = exports.execa = exports.disableTypeChecksHelper = exports.checkerConcurrencyTokens = exports.checkerFactory = exports.checkerPool = void 0;
exports.checkerPool = 'checkerPool';
exports.checkerFactory = 'checkerFactory';
exports.checkerConcurrencyTokens = 'checkerConcurrencyTokens';
exports.disableTypeChecksHelper = 'disableTypeChecksHelper';
exports.execa = 'execa';
exports.cliOptions = 'cliOptions';
exports.configReader = 'configReader';
exports.configOptionsApplier = 'configOptionsApplier'; // TODO: Remove if unused post mutation switching
exports.inputFiles = 'inputFiles';
exports.dryRunResult = 'dryRunResult';
exports.transpiledFiles = 'transpiledFiles'; // TODO: Remove if unused post mutation switching
exports.files = 'files';
exports.mutants = 'mutants';
exports.mutantsWithTestCoverage = 'mutantsWithTestCoverage';
exports.mutantTranspileScheduler = 'mutantTranspileScheduler'; // TODO: Remove if unused post mutation switching
exports.temporaryDirectory = 'temporaryDirectory';
exports.timer = 'timer';
exports.timeOverheadMS = 'timeOverheadMS';
exports.loggingContext = 'loggingContext';
exports.mutationTestReportHelper = 'mutationTestReportHelper';
exports.sandbox = 'sandbox';
exports.sandboxWorkingDirectory = 'sandboxWorkingDirectory'; // TODO: Remove if unused post mutation switching
exports.concurrencyTokenProvider = 'concurrencyTokenProvider';
exports.testRunnerFactory = 'testRunnerFactory';
exports.testRunnerPool = 'testRunnerPool';
exports.testRunnerConcurrencyTokens = 'testRunnerConcurrencyTokens';
exports.reporter = 'reporter';
exports.pluginKind = 'pluginKind';
exports.pluginDescriptors = 'pluginDescriptors';
exports.pluginCreatorReporter = 'pluginCreatorReporter';
exports.pluginCreatorChecker = 'pluginCreatorChecker';
exports.pluginCreatorTestRunner = 'pluginCreatorTestRunner';
exports.validationSchema = 'validationSchema';
exports.optionsValidator = 'optionsValidator';
