// @stryker-mutator/util/src/string-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.escapeRegExp = exports.escapeRegExpLiteral = exports.PropertyPathBuilder = exports.propertyPath = exports.normalizeWhitespaces = void 0;
/**
 * Consolidates multiple consecutive white spaces into a single space.
 * @param str The string to be normalized
 */
function normalizeWhitespaces(str) {
    return str.replace(/\s+/g, ' ').trim();
}
exports.normalizeWhitespaces = normalizeWhitespaces;
/**
 * Given a base type, allows type safe access to the name of a property.
 * @param prop The property name
 */
function propertyPath(prop) {
    return prop.toString();
}
exports.propertyPath = propertyPath;
/**
 * A helper class to allow you to get type safe access to the name of a deep property of `T`
 * @example
 * ```ts
 * PropertyPathBuilder<StrykerOptions>('warnings').prop('unknownOptions').build()
 * ```
 */
class PropertyPathBuilder {
    constructor(pathSoFar) {
        this.pathSoFar = pathSoFar;
    }
    prop(prop) {
        return new PropertyPathBuilder([...this.pathSoFar, prop.toString()]);
    }
    /**
     * Build the (deep) path to the property name
     */
    build() {
        return this.pathSoFar.join('.');
    }
    /**
     * Creates a new `PropertyPathBuilder` for type T
     */
    static create() {
        return new PropertyPathBuilder([]);
    }
    toString() {
        return this.build();
    }
}
exports.PropertyPathBuilder = PropertyPathBuilder;
/**
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions#Escaping
 */
function escapeRegExpLiteral(input) {
    return input.replace(/[.*+\-?^${}()|[\]\\/]/g, '\\$&'); // $& means the whole matched string
}
exports.escapeRegExpLiteral = escapeRegExpLiteral;
/**
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions#Escaping
 */
function escapeRegExp(input) {
    return input.replace(/[.*+\-?^${}()|[\]\\]/g, '\\$&'); // $& means the whole matched string
}
exports.escapeRegExp = escapeRegExp;
